src/proc/color-formats-converter.cpp

```diff
@@ -354,6 +358,207 @@ namespace librealsense
+    // The first pixel is (Y0, U0, V0), second pixel is (Y1, U0, V0)
+    // The first pixel in the second line is (Yw, U0, V0) second pixel in second line is (Yw+1, U0, V0)
+    // The third pixel in second line is (Yw+2, U1, V1)
+    template<rs2_format FORMAT> void unpack_nv12i(byte* const d[], const byte* s, int width, int height, int actual_size)
+    {
+        auto n = width * height;
+        assert(n % 16 == 0); // All currently supported color resolutions are multiples of 16 pixels. Could easily extend support to other resolutions by copying final n<16 pixels into a zero-padded buffer and recursively calling self for final iteration.
+
+        auto src = reinterpret_cast<const uint8_t*>(s);
+        auto dst = reinterpret_cast<uint8_t*>(d[0]);
+
+        auto src_height = height * 12 >> 3;
+
+        if (FORMAT == RS2_FORMAT_Y8)
+        {
+            // TODO REMI - check height is correct here (shall be equal to 1.5 * height - for height = 480, shall be 720)
+            for (int k = 0; k < src_height; k += 3)
+            {
+                // fill the y_buffer and uv_buffer
+                // while y is on 2 lines, and uv on the third line
+                auto start_of_y = src + k * width;
+                auto end_of_y = start_of_y + 2 * width;
+
+                std::vector<uint8_t> y_buffer(start_of_y, end_of_y);
+                librealsense::copy(dst, y_buffer.data(), 2 * width);
+                dst += 2 * width;
+            }
+            return;
+        }
+        if (FORMAT == RS2_FORMAT_Y16)
+        {
+            for (int k = 0; k < src_height; k += 3)
+            {
+                // fill the y_buffer and uv_buffer
+                // while y is on 2 lines, and uv on the third line
+                auto start_of_y = src + k * width;
+                auto end_of_y = start_of_y + 2 * width;
+                auto start_of_uv = end_of_y;
+                auto end_of_uv = start_of_uv + width;
+
+                std::vector<uint8_t> y_buffer(start_of_y, end_of_y);
+
+                for (int pix = 0; pix < 2 * width; pix += 16)
+                {
+                    uint16_t y[16];
+                    for (int i = 0; i < 16; ++i)
+                    {
+                        y[i] = y_buffer[pix + i];
+                    }
+                    librealsense::copy(dst, &y[0], 16);
+                    dst += 16;
+                }
+            }
+            return;
+        }
+        
+        for (int k = 0; k < src_height; k += 3)
+        {
+            // fill the y_buffer and uv_buffer
+            // while y is on 2 lines, and uv on the third line
+            auto start_of_y = src + k * width;
+            auto start_of_second_line = start_of_y + width;
+            auto end_of_y = start_of_second_line + width;
+            auto start_of_uv = end_of_y;
+            auto end_of_uv = start_of_uv + width;
+
+            std::vector<uint8_t> y_buffer_first_line(start_of_y, start_of_second_line);
```

Comment:
I know it's more readable, but we should aim for removing all unneeded copies.
This function will run many times for each frame.
We can use the original input all the way

Reply:
Done